### DISCUSSION TITLE
More fexible Graphite/Loki query mapping

### DISCUSSION THREAD

**Discussion by Dieterbe** [grafana/grafana#42601] [Observability]
The Graphite/Loki query mapping (https://github.com/grafana/grafana/pull/33405) is a really neat future, but I run into some limations with it

Let's say I have a Graphite query like `metrictank.stats.us-east2-854-ops.mt-query-854-ops-45w8b.api.foo.bar`
this queries stats from a pod that emits loki logs with the following labels:
```
app	metrictank
cluster	hm-us-east2
color	b
container	metrictank
hosted_metrics	1
instance	ops
job	metrictank/metrictank
namespace	metrictank
org	854
pod	mt-query-854-ops-45w8b
pod_template_hash	6bf75b545b
role	query
```

currently (and assuming #34003 gets merged), i can only use a mapping like `metrictank.stats.*.(pod)`: i can only extract the pod label because it's the only exact match.
I suggest the following improvements:

1) ability to parse parts out of nodes. Note how the word `ops` appears in the graphite query (but not as a full word). I'ld like to extract this so that I can set the `instance` label with it. it improves loki index performance to have multiple matchers, even if they result in the same log streams. But more importantly. if my query had `mt-query.*` in it, it would match many pods also from different clusters and instances, which I don't want
2) ability to assign static prefixes to label values. note how the cluster is "hm-us-east2", but in the graphite query it only says "us-east2"
3) i may want to set extra labels that are static, e.g. if the pattern maches (and regardless of the parsed values), also set `namespace=metrictank` and/or `hosted_metrics=1` in the query. (I could envision how some people also may have 1 label whose value should be a string concatenation of multiple graphite nodes or parts parsed out of nodes)

To be fair, only the first one is a real need for me. I would probably be just fine without 2/3, i just mention them because they seem like related problems

**Comment by ifrost** [comment]
Maybe for more complex scenarios we could allow switching to pseudo PromQL/LokiQL like syntax with special functions to create label values? For example:

searched metric is: `metrictank.stats.us-east2-854-ops.mt-query-854-ops-45w8b.api.foo.bar`

so we have these nodes:
node 1: `metrictank`
node 2: `stats`
node 3: `us-east2-854-ops`
node 4: `mt-query-854-ops-45w8b`

and now the mapping is defined as

```php
metrictank.stats.*.mt-query*.api.foo.bar
{ 
  app=node(1), 
  cluster=split(3, "-", "hm-$0-$1"), 
  org=split(3, "-", "$2"), 
  instance=split(3, "-", "$3"), 
  pod=node(4),
  myLabel="myValue" }
}
```

where:
- `metrictank.stats.*.mt-query*.api.foo.bar` is Graphite query to match against before extracting labels
- `node(index)` - returns the node at given index as the value
- `split(index, character, pattern)` - returns value created by splitting the node at given index and constructing it from split parts using the pattern with $0, $1 as indices.

that way it would create a Loki Query:

```
{ 
  app="metrictank", 
  cluster="hm-us-east2", 
  org="854", 
  instance="ops", 
  pod="mt-query-854-ops-45w8b", 
  myLabel="myValue" 
}
```

We could also go with regular expressions, but that can get very complex quickly with no easy way to debug it. 🤪

**Comment by Dieterbe** [comment]
Note that in graphite, nodes are typically counted starting from 0 (e.g. in `aliasByNode`)
I like your idea, although it is on the verbose side, but the components are digestible.
With regex (or similar) we make it a much more compact pattern that tells us how to parse a pattern and using named groups we can name the labels that we extract' but it wouldn't allow for e.g. the static prefix, and setting extra hardcoded labels. it would also look more complicated (if we use regex, anyway)

I suspect that most people using the feature would probably only have a couple of rules, so there isn't much cost to them being more verbose.

**Comment by ifrost** [comment]
To support static labels with regex we could split mapping into 2 inputs - one for capturing values with regex, and one for creating Loki query, e.g.:

capturing values with named groups: `/(?<app>metrictank)/.stats./^(?<cluster>\w{2}-\w+)-(?<org>\d+)-(?<instance>\w+)$/./(?<pod>.*)/`
creating Loki query: `{app="${app}", cluster="hm-${cluster}", pod="${pod}", org="${org}", instance="${instance}", static="blah"}`

In this example each node has a separate regex but maybe a single regex for entire path could work, too? Not sure how it'd play with graphite glob pattern, e.g. what if for the mapping above the query was `metrictank.stats.us-{west,east}2.*` 🤔

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
